src/generic/Matrix.jl

```diff
@@ -157,63 +159,44 @@ end
 #
 ###############################################################################
 
+# create a zero matrix
 function (a::MatSpace{T})() where {T <: NCRingElement}
-   R = base_ring(a)
-   entries = Array{T}(undef, a.nrows, a.ncols)
-   for i = 1:a.nrows
-      for j = 1:a.ncols
-         entries[i, j] = zero(R)
-      end
-   end
-   z = MatSpaceElem{T}(R, entries)
-   return z
+   return dense_matrix_type(T)(base_ring(a), nrows(a), ncols(a))
```

Comment:
So the idea is that every matrix type should provide at least one constructor with signature `R::NCRing, r::Int, c::Int` which produces a zero matrix, and we define everything else in terms of that. (Of course additional such methods can be installed for specific matrix types as an optimization; but the generic code always first creates a zero matrix and then adjusts its entries as needed)